service.go

```diff
@@ -0,0 +1,174 @@
+	{
+		config := memory.DefaultConfig()
+		newStorage, err = memory.New(config)
+		if err != nil {
+			panic(err)
+		}
+	}
+
+	return Config{
+		// Dependencies.
+		Logger:  newLogger,
+		Storage: newStorage,
+	}
+}
+
+// New creates a new configured ipam service.
+func New(config Config) (*Service, error) {
+	// Dependencies.
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "logger must not be empty")
+	}
+	if config.Storage == nil {
+		return nil, microerror.Maskf(invalidConfigError, "storage must not be empty")
+	}
+
+	// Settings.
+	if config.Network == nil {
+		return nil, microerror.Maskf(invalidConfigError, "network must not be empty")
+	}
+
+	newService := &Service{
+		// Dependencies.
+		logger:  config.Logger,
+		storage: config.Storage,
+
+		// Settings.
+		network: *config.Network,
+	}
+
+	return newService, nil
+}
+
+type Service struct {
+	// Dependencies.
+	logger  micrologger.Logger
+	storage microstorage.Storage
+
+	// Settings.
+	network net.IPNet
+}
+
+// key returns a storage key for a given network.
+func key(network net.IPNet) string {
+	return fmt.Sprintf(
+		IPAMSubnetStorageKeyFormat,
+		strings.Replace(network.String(), "/", "-", -1),
+	)
+}
+
+// listSubnets retrieves the stored subnets from storage and returns them.
+func (s *Service) listSubnets(ctx context.Context) ([]net.IPNet, error) {
+	s.logger.Log("info", "listing subnets")
+
+	existingSubnetStrings, err := s.storage.List(ctx, IPAMSubnetStorageKey)
+	if err != nil && !microstorage.IsNotFound(err) {
+		return nil, microerror.Mask(err)
+	}
+
+	existingSubnets := []net.IPNet{}
+	for _, existingSubnetString := range existingSubnetStrings {
+		// Storage returns the relative key with List, not the values.
+		// Instead of then requesting each value, we revert the key to a valid
+		// CIDR string.
+		existingSubnetString = strings.Replace(existingSubnetString, "-", "/", -1)
+
+		_, existingSubnet, err := net.ParseCIDR(existingSubnetString)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+		existingSubnets = append(existingSubnets, *existingSubnet)
+	}
+
+	subnetCounter.Set(float64(len(existingSubnets)))
+
+	return existingSubnets, nil
+}
+
+// NewSubnet returns the next available subnet, of the configured size,
+// from the configured network.
+func (s *Service) NewSubnet(mask net.IPMask) (net.IPNet, error) {
+	s.logger.Log("info", "creating new subnet")
+	defer updateMetrics("create", time.Now())
+
+	ctx := context.Background()
+
+	existingSubnets, err := s.listSubnets(ctx)
+	if err != nil {
+		return net.IPNet{}, microerror.Mask(err)
+	}
+
+	s.logger.Log("info", "computing next subnet")
+	subnet, err := Free(s.network, mask, existingSubnets)
+	if err != nil {
+		return net.IPNet{}, microerror.Mask(err)
+	}
+
+	s.logger.Log("info", "putting subnet", "subnet", subnet.String())
+	if err := s.storage.Put(ctx, key(subnet), subnet.String()); err != nil {
+		return net.IPNet{}, microerror.Mask(err)
+	}
```

Comment:
I wonder if there is a solution to the problem when the subnet is stored (so not free), but the calling code fails, and not uses it at all.

Probably more metrics.